source/common/http/conn_manager_impl.cc

```diff
@@ -564,28 +564,35 @@ void ConnectionManagerImpl::ActiveStream::traceRequest() {
   if (cached_route_.value() && cached_route_.value()->decorator()) {
     cached_route_.value()->decorator()->apply(*active_span_);
 
+    // Cache decorated operation.
+    if (!cached_route_.value()->decorator()->getOperation().empty()) {
+      decorated_operation_ = &cached_route_.value()->decorator()->getOperation();
+    }
+  }
+
+  if (connection_manager_.config_.tracingConfig()->operation_name_ ==
+      Tracing::OperationName::Egress) {
```

Comment:
I think we need to discuss https://github.com/envoyproxy/envoy/pull/1931 in the context of this PR also. I'm not completely sure what the best way to proceed is and it may effect this PR also. Basically, middle proxies perform both ingress and egress effectively. Currently, at Lyft we set our edge proxies to ingress direction (basically https://github.com/envoyproxy/envoy/blob/master/configs/envoy_front_proxy.template.json#L30). 

I could see in general wanting to start router spans whenever direction is set to egress. But then we would need a new ingress/egress mode. Or do you think we should add a new config element at the HTTP connection manager tracing config level such as "start_child_spans_for_nested_operations" ?

Reply:
cc @goaway

Reply:
@mattklein123 I guess one question would be - would you expect a middle proxy to create `client` and `server` spans?

If not, then possibly adding an `IngressAndEgress` operation_name value in the tracing config might work, with suitable guards on all if/else clauses to ensure the client/server spans are only created if either `Ingress` or `Egress` operation_name is specified. Or it may just work if no operation_name is specified (so no need for new value).

Other thing to consider - currently the `client` and `server` spans are created in the same place (conn_manager_impl traceRequest) - but possibly only the `server` span should be created here when `Ingress`, and the `client` span should be created for `Egress` in the router?

Reply:
My suspicion is that most users would expect a middle proxy to create an ingress span (for systems where that matters like Zipkin).

I'm still struggling to decide if it's worth adding IngressEgress and potentially driving the router logic via that. 

> Other thing to consider - currently the client and server spans are created in the same place (conn_manager_impl traceRequest) - but possibly only the server span should be created here when Ingress, and the client span should be created for Egress in the router?

FWIW I don't think so, because even in pure egress, there may be other filters before router where we want to have a parent to attach to.

The more I think about this the more I think that we should either go with my original solution which is to have a config property on the router (since conceivably you want this behavior on ingress also) or a top level tracing config on the HTTP connection manager such as "start_child_spans_for_nested_ops". Thoughts?

Reply:
Alright I chatted with @goaway about this. This is the current thinking:

- Ingress: Start ingress span in HTTP connection manager
- IngressEgress: Start ingress span in HTTP connection manager, start egress span in router.
- Egress: Start egress span in HTTP connection manager, and egress span child in router. (So basically it's EgressEgress).

The issue with ^ is that I guarantee someone is going to come along and say: "Why is Envoy starting two spans on egress. I don't use retries. It looks ugly." I could go either way on this.

Reply:
I don't think the Egress approach would work, as this would create a `client` span in the connection manager, and then another child span - which would then have its trace context propagated to the invoked service (proxy) which would create the `server` span - so there would be a child span between the `client` and `server` span in the trace hierarchy.

This is why I was thinking that the Egress (client) span would need to be created in the router - and the Egress connection manager would just create a (component) span acting as the parent for the potential multiple retried router (client) spans.

As you mentioned, maybe best to go back to your original solution for Ingress only for now? So provides additional (optional on config) details of how the proxy is handling the ingress request. Can then consider the Egress case separately?

Reply:
@objectiser is there any chance you could hop into Envoy slack and we could chat about this? I'm honestly pretty confused at this point. Part of the problem is I don't fully understand the implications of ingress/egress as it relates to Zipkin as LS does not have this concept. I'm also not familiar with concepts like "component" span.

I think we do want to be able to create a child span in the router for egress at Lyft, as that will allow us to see retries.